# Exploración de Datos — Predicción de Diabetes
**Proyecto Integrado 3 — Actividad 1**

**Dataset:** Pima Indians Diabetes Database (Kaggle / UCI)  
**Objetivo:** Realizar una exploración exhaustiva del dataset usando Pandas Profiling para identificar la calidad de los datos, distribuciones, correlaciones y valores atípicos.

## 1. Instalación de Dependencias

In [1]:
!pip install ydata-profiling pandas

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 400.8/400.8 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.7/296.7 kB 19.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 59.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 679.7/679.7 kB 36.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 105.4/105.4 kB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.0/72.0 kB 4.7 MB/s eta 0:00:00


## 2. Carga del Dataset

In [2]:
import pandas as pd
from ydata_profiling import ProfileReport

df = pd.read_csv('/content/diabetes.csv')

df = df.rename(columns={
    'Pregnancies': 'Embarazos',
    'Glucose': 'Glucosa',
    'BloodPressure': 'PresionArterial',
    'SkinThickness': 'GrosorPiel',
    'Insulin': 'Insulina',
    'BMI': 'IndiceMasaCorporal',
    'DiabetesPedigreeFunction': 'FuncionPedigree',
    'Age': 'Edad',
    'Outcome': 'Diagnostico'
})

print(f'Dimensiones del dataset: {df.shape[0]} filas x {df.shape[1]} columnas')
df.head(10)

Dimensiones del dataset: 768 filas x 9 columnas


/tmp/ipykernel_29462/1457277131.py:2: DeprecationWarning: 
    `import ydata_profiling` is deprecated and will not receive more updates. 
    Please install fg-data-profiling via `pip install fg-data-profiling` and use `import data_profiling` instead.
    
  from ydata_profiling import ProfileReport


,Embarazos,Glucosa,PresionArterial,GrosorPiel,Insulina,IndiceMasaCorporal,FuncionPedigree,Edad,Diagnostico
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1
5,5,116,74,0,0,25.6,0.201,30,0
6,3,78,50,32,88,31.0,0.248,26,1
7,10,115,0,0,0,35.3,0.134,29,0
8,2,197,70,45,543,30.5,0.158,53,1
9,8,125,96,0,0,0.0,0.232,54,1


## 3. Exploración Básica

In [3]:
print('=== Tipos de datos ===')
print(df.dtypes)
print()
print('=== Valores nulos por columna ===')
print(df.isnull().sum())
print()
print('=== Registros duplicados ===')
print(f'{df.duplicated().sum()} duplicados encontrados')

=== Tipos de datos ===
Embarazos               int64
Glucosa                 int64
PresionArterial         int64
GrosorPiel              int64
Insulina                int64
IndiceMasaCorporal    float64
FuncionPedigree       float64
Edad                    int64
Diagnostico             int64
dtype: object

=== Valores nulos por columna ===
Embarazos             0
Glucosa               0
PresionArterial       0
GrosorPiel            0
Insulina              0
IndiceMasaCorporal    0
FuncionPedigree       0
Edad                  0
Diagnostico           0
dtype: int64

=== Registros duplicados ===
0 duplicados encontrados


In [4]:
print('=== Estadísticas descriptivas ===')
df.describe()

=== Estadísticas descriptivas ===


,Embarazos,Glucosa,PresionArterial,GrosorPiel,Insulina,IndiceMasaCorporal,FuncionPedigree,Edad,Diagnostico
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


### Nota sobre valores 0 como nulos implícitos

En este dataset, las columnas `Glucosa`, `PresionArterial`, `GrosorPiel`, `Insulina` e `IndiceMasaCorporal` contienen valores `0` que **no son fisiológicamente posibles**. Estos representan datos faltantes codificados como cero, y deben tratarse como nulos en etapas posteriores de preprocesamiento.

In [5]:
cols_con_nulos_implicitos = ['Glucosa', 'PresionArterial', 'GrosorPiel', 'Insulina', 'IndiceMasaCorporal']
print('=== Conteo de ceros (nulos implícitos) ===')
for col in cols_con_nulos_implicitos:
    zeros = (df[col] == 0).sum()
    pct = zeros / len(df) * 100
    print(f'  {col}: {zeros} ceros ({pct:.1f}%)')

=== Conteo de ceros (nulos implícitos) ===
  Glucosa: 5 ceros (0.7%)
  PresionArterial: 35 ceros (4.6%)
  GrosorPiel: 227 ceros (29.6%)
  Insulina: 374 ceros (48.7%)
  IndiceMasaCorporal: 11 ceros (1.4%)


## 4. Balance de la Variable Objetivo

In [6]:
print('=== Distribución de Diagnostico ===')
counts = df['Diagnostico'].value_counts()
pcts = df['Diagnostico'].value_counts(normalize=True) * 100
print(f'  No diabético (0): {counts[0]} registros ({pcts[0]:.1f}%)')
print(f'  Diabético    (1): {counts[1]} registros ({pcts[1]:.1f}%)')

=== Distribución de Diagnostico ===
  No diabético (0): 500 registros (65.1%)
  Diabético    (1): 268 registros (34.9%)


## 5. Reporte Completo con Pandas Profiling

El siguiente reporte genera de forma automática:
- Distribución de cada variable (histogramas)
- Detección de valores faltantes y outliers
- Matriz de correlaciones
- Interacciones entre variables
- Estadísticas descriptivas completas

In [7]:
profile = ProfileReport(
    df,
    title='Exploración: Pima Indians Diabetes Dataset',
    explorative=True
)
profile.to_notebook_iframe()

Output hidden; open in https://colab.research.google.com to view.

In [8]:
profile.to_file('diabetes_report.html')
print('Reporte guardado en: diabetes_report.html')

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

Reporte guardado en: diabetes_report.html


## 6. Análisis de Correlaciones con la Variable Objetivo

In [9]:
print('=== Correlación de cada variable con Diagnostico ===')
correlaciones = df.corr()['Diagnostico'].drop('Diagnostico').sort_values(ascending=False)
print(correlaciones.to_string())

=== Correlación de cada variable con Diagnostico ===
Glucosa               0.466581
IndiceMasaCorporal    0.292695
Edad                  0.238356
Embarazos             0.221898
FuncionPedigree       0.173844
Insulina              0.130548
GrosorPiel            0.074752
PresionArterial       0.065068


## 7. Conclusiones y Hallazgos

### 7.1 Calidad del Dataset

El dataset tiene 768 registros y 9 variables, sin valores nulos explícitos ni filas duplicadas. Sin embargo, el reporte de Pandas Profiling detectó **9 alertas** relacionadas con ceros que no tienen sentido clínico:

| Variable | Ceros | Porcentaje |
|---|---|---|
| Insulina | 374 | 48.7% |
| GrosorPiel | 227 | 29.6% |
| Embarazos | 111 | 14.5% |
| PresionArterial | 35 | 4.6% |
| IndiceMasaCorporal | 11 | 1.4% |

La variable con mayor problema es `Insulina`, donde casi la mitad de los registros tienen valor 0. Estos valores deben tratarse como datos faltantes en etapas posteriores.

### 7.2 Correlaciones con el Diagnóstico

Las variables con mayor correlación con `Diagnostico` fueron:

1. **Glucosa (0.47)** — es la variable más relacionada con el diagnóstico de diabetes, lo cual tiene sentido clínico directo.
2. **IndiceMasaCorporal (0.29)** — el sobrepeso es un factor de riesgo conocido.
3. **Edad (0.24)** — a mayor edad, mayor probabilidad de diagnóstico positivo.
4. **Embarazos (0.22)** — guarda relación con la diabetes gestacional.
5. **FuncionPedigree (0.17)** — refleja la predisposición genética familiar.

Las variables `GrosorPiel` (0.07) y `PresionArterial` (0.07) tienen muy baja correlación con el diagnóstico.

### 7.3 Correlaciones entre Variables

El reporte también detectó correlación alta entre pares de variables independientes:
- **Edad ↔ Embarazos**: tiene sentido, ya que a mayor edad hay más probabilidad de haber tenido más embarazos.
- **GrosorPiel ↔ Insulina**: ambas variables miden indicadores metabólicos relacionados entre sí.

Esto puede ser relevante si en el futuro se aplican modelos sensibles a la multicolinealidad.

### 7.4 Próximos Pasos

1. Reemplazar los ceros de `Insulina`, `GrosorPiel`, `PresionArterial` e `IndiceMasaCorporal` con la mediana agrupada por `Diagnostico`.
2. Revisar outliers en `Insulina` (valores > 600).
3. Construir un modelo de clasificación usando las variables con mayor correlación como punto de partida.